tools/build-pypi-artifacts.py

```diff
@@ -0,0 +1,253 @@
+                            stdout=subprocess.PIPE)
+    status, _ = proc.communicate()
+    status = status.decode("ascii", "replace")
+    if status.endswith('\n'):
+        commit_id = status[:-1]
+
+    assert len(commit_id) == 40
+    return commit_id
+
+
+def commit_changed_files():
+    proc = subprocess.Popen(['git', 'commit', '-a', '-m',
+                             '"Commit for PyPI sdist (by build-pypi-artifacts.py)"'],
+                            stdout=subprocess.PIPE)
+    status, _ = proc.communicate()
+    status = status.decode("ascii", "replace")
+    if not ("2 files changed" in status and "rename LICENSES" in status):
+        print(f"{status}\n")
+        raise RuntimeError("Expected changes to 3 files: pyproject.toml, "
+                           "LICENSE.txt, and LICENSES_bundled.txt")
+
+
+def reset_to_previous_commit(commit_id):
+    proc = subprocess.Popen(['git', 'reset', '--hard', commit_id],
+                            stdout=subprocess.PIPE)
+    status, _ = proc.communicate()
+    status = status.decode("ascii", "replace")
+    if not "HEAD is now at" in status:
+        print(f"{status}\n")
+        raise RuntimeError("Resetting to original commit may have gone wrong")
+
+
+def generate_sdist():
+    proc = subprocess.run([sys.executable, '-m', 'build', '--sdist',
+                           '--no-isolation', '--skip-dependency-check'])
+    distdir = repo_root_dir / 'dist'
+    if not len(list(distdir.glob('scipy-*.tar.gz'))) == 1:
+        raise RuntimeError("sdist creating not succesful!")
+
+
+def create_sdist_for_pypi():
+    check_submodules()
+    check_not_dirty()
+    commit_id = get_current_commit_id()
+
+    concat_license_files()
+    modify_pyproject_toml()
+
+    commit_changed_files()
+    generate_sdist()
+    reset_to_previous_commit(commit_id)
+
+
+def create_wheel_for_pypi(skip_check=False):
+    """
+    Create a wheel with the appropriate wheel-specific license info (for
+    vendored libraries) and `_distributor_init.py` file.
+    """
+    # TODO: adapt for what is going on in https://github.com/MacPython/scipy-wheels
+    #       - grab the correct build of OpenBLAS (or keep separate?)
+    #       - patch_code.sh changes need to live here (double check and
+    #         document what mesonpy does with vendoring)
+    #       - patch_numpy.sh isn't needed
+    #       - set MACOSX_DEPLOYMENT_TARGET in environment here
+    #       - unclear whether we have a use for setting PYTHONFAULTHANDLER=1
+    #       - verify_init.py is likely good to incorporate here
+    raise NotImplementedError('TODO')
+
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(
+        description=textwrap.indent(
+            textwrap.dedent(
+                """
+                Tool to build SciPy release artifacts (sdist, wheels) for PyPI.
+
+                This tool is needed for releases because we need to take
+                PyPI-specific actions like amend license files and
```

Comment:
Ah yes, there's overlap. The way you have it in gh-15925 is good for `setup.py`, it won't work well with `meson-python` in its current form though, because it picks up only committed changes so you cannot apply any patches - hence my conclusion that that needs changing.

Reply:
I think we're going to change the approach here, so your gh-15925 changes are fine as they are now. I just needed the changes in this PR to actually figure out how problematic the current `meson-python` behavior is.

Reply:
> because it picks up only committed changes so you cannot apply any patches

That's super-confusing, and would indeed be great if that'd change. (a drive-by comment)